pkg/products/monitoringspec/reconciler.go

```diff
@@ -0,0 +1,388 @@
+	if err != nil {
+		return integreatlyv1alpha1.PhaseFailed, err
+	}
+
+	ls, err := labels.Parse(labelSelector)
+	if err != nil {
+		return integreatlyv1alpha1.PhaseFailed, err
+	}
+	opts := &k8sclient.ListOptions{
+		LabelSelector: ls,
+	}
+
+	//Get the list of namespaces with the given label selector
+	namespaces := &corev1.NamespaceList{}
+	err = serverClient.List(ctx, namespaces, opts)
+	if err != nil {
+		return integreatlyv1alpha1.PhaseFailed, err
+	}
+
+	for _, ns := range namespaces.Items {
+		//Get list of service monitors in each name space
+		listOpts := []k8sclient.ListOption{
+			k8sclient.InNamespace(ns.Name),
+		}
+		serviceMonitors := &v1.ServiceMonitorList{}
+		err := serverClient.List(ctx, serviceMonitors, listOpts...)
+		if err != nil {
+			return integreatlyv1alpha1.PhaseFailed, err
+		}
+		for _, sm := range serviceMonitors.Items {
+			//Create a copy of service monitors in the monitoring namespace
+			//Create the corresponding rolebindings at each of the service namespace
+			key := sm.Namespace + `-` + sm.Name
+			delete(monSermonMap, key) // Servicemonitor exists, remove it from the local map
+			err := r.reconcileServiceMonitor(ctx, serverClient, sm)
+			if err != nil {
+				return integreatlyv1alpha1.PhaseFailed, err
+			}
+		}
+	}
+
+	//Clean-up the stale service monitors and rolebindings if any
+	if len(monSermonMap) > 0 {
+		for _, sm := range monSermonMap {
+			//Remove servicemonitor
+			err = r.removeServicemonitor(ctx, serverClient, sm.Namespace, sm.Name)
+			if err != nil {
+				return integreatlyv1alpha1.PhaseFailed, err
+			}
+			//Remove rolebindings
+			for _, namespace := range sm.Spec.NamespaceSelector.MatchNames {
+				err := r.removeRoleBinding(ctx, serverClient, namespace, roleBindingName)
+				if err != nil {
+					return integreatlyv1alpha1.PhaseFailed, err
+				}
+			}
+		}
+	}
+	return integreatlyv1alpha1.PhaseCompleted, err
+}
+
+func (r *Reconciler) reconcileServiceMonitor(ctx context.Context,
+	serverClient k8sclient.Client, serviceMonitor *v1.ServiceMonitor) (err error) {
+
+	if serviceMonitor.Spec.NamespaceSelector.Any {
+		logrus.Warnf("servicemonitor : %s cannot be copied to %s namespace. Namespace selector has been set to any",
+			serviceMonitor.Name, r.Config.GetNamespace())
+		return nil
+	}
+	sm := &v1.ServiceMonitor{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceMonitor.Namespace + `-` + serviceMonitor.Name,
+			Namespace: r.Config.GetNamespace(),
+		},
+	}
+	opRes, err := controllerutil.CreateOrUpdate(ctx, serverClient, sm, func() error {
+		// Check if the servicemonitor has no  namespace selectors defined,
+		// if not add the namespace
+		sm.Spec = serviceMonitor.Spec
+		if len(sm.Spec.NamespaceSelector.MatchNames) == 0 {
+			sm.Spec.NamespaceSelector.MatchNames = []string{serviceMonitor.Namespace}
+		}
+		//Add all the original labels and append cloned servicemonitor label
+		sm.Labels = serviceMonitor.Labels
+		if len(sm.Labels) == 0 {
```

Comment:
Should this be a `nil` check rather than a length check?

Reply:
Empty map, length can be checked in golang.  ```package main

import "fmt"

func main() {
	var m map[string]string
	fmt.Println(len(m))
	if m == nil {
		fmt.Println("map is nil")
	}
}```  output: ```

1

package main

2

​

3

import "fmt"

4

​

5

func main() {

6

    var m map[string]string

7

    fmt.Println(len(m))

8

    if m == nil {

9

        fmt.Println("map is nil")

10

    }

11

}

12

​

0
map is nil```